lib/extensions.js

```diff
@@ -268,8 +270,30 @@ module.exports = function (realmConstructor) {
           }),
         );
 
-        // Return wrapped promises, allowing the users to control them.
-        let openPromise = Promise.race(openPromises);
+        // Return wrapped promises, allowing the users to control them. Once one of the
+        // `openPromise`s has resolved, we may need to wait for initial subscriptions
+        // (if any) to be synchronised, so we return a chained promise to do this.
+        let openPromise = Promise.race(openPromises).then((realm) => {
+          const { initialSubscriptions } = config.sync;
+
+          // If `initialSubscriptions` was not specified, return the Realm immediately
+          if (!initialSubscriptions) {
+            return realm;
+          }
+
+          // If an update has been run by C++ (which performs all the validation
+          // and runs the actual update function, see `handle_initial_subscriptions`
+          // in `js_realm.hpp`), we need to return a promise which waits for the
+          // new subscriptions to be fully synchronised, then returns the Realm.
+          if (initialSubscriptions.rerunOnStartup || !realmExists) {
+            return realm.subscriptions.waitForSynchronization().then(() => {
+              return realm;
+            });
+          } else {
+            return realm;
+          }
+        });
```

Comment:
If you made the chained promise callback an async function, you could simplify the logic at the bottom of the function body. Just a thought.

```suggestion
        let openPromise = Promise.race(openPromises).then(async (realm) => {
          const { initialSubscriptions } = config.sync;

          // If `initialSubscriptions` was not specified, return the Realm immediately
          if (!initialSubscriptions) {
            return realm;
          }

          // If an update has been run by C++ (which performs all the validation
          // and runs the actual update function, see `handle_initial_subscriptions`
          // in `js_realm.hpp`), we need to return a promise which waits for the
          // new subscriptions to be fully synchronised, then returns the Realm.
          if (initialSubscriptions.rerunOnStartup || !realmExists) {
            await realm.subscriptions.waitForSynchronization();
          }
          return realm;
        });
```